[SourceKit] Support building sourcekitd without building swift-syntax

ahoppen opened the pull request:
This allows building sourcekitd and swift-refactor with `SWIFT_BUILD_SWIFT_SYNTAX=NO`. In these builds, the `relatedidents` and `find-syntactic-rename-ranges` requests will always return an error.

bnbarham commented:
> Do you run refactoring tests for Android? If so, I think we need to disable those as well.

They would all need a `swift_swift_parser` guard now. There's more than just this too, `swiftRefactoring` isn't built either and thus sourcekit will also fail to build.

What we probably need to do here is to have `swiftRefactoring` only depend on `swiftIDEUtilsBridging` if swift-syntax was built, then `#if` any sections of `swiftRefactoring` that depend on `swiftIDEUtilsBridging`. And maybe a warning in sourcekit to say that rename is unavailable without early swift-syntax.

ahoppen commented:
@finagolfin Do you use sourcekitd and Swift tools on Android? If not, you could try setting `SWIFT_BUILD_SOURCEKIT=NO` and `SWIFT_INCLUDE_TOOLS=NO` to just not build sourcekitd and `swift-refactor` (+ the other tools in `swift/tools`)

finagolfin commented:
> Do you run refactoring tests for Android?

Yes, it appears [they are normally run on the community Android CI](https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-16.04-android-arm64/lastSuccessfulBuild/consoleText), which I don't control but help monitor.

> Do you use sourcekitd and Swift tools on Android?

Yes, I do. In fact, I even [distribute sourcekit-lsp natively on Android](https://forums.swift.org/t/cross-compilation-sdks-for-android/49101/5), and have used it with an LSP-capable editor in the Termux terminal app for Android.

In any case, this issue has nothing to do with Android, but is about making the build and tests work with `SWIFT_BUILD_SWIFT_SYNTAX` turned off on _any_ platform. I know the plan is to eventually make that flag a requirement, which is why every official CI now has a host Swift compiler and doesn't bother testing without it, but the community Android CI, which only runs on Ubuntu linux, doesn't install a prebuilt Swift compiler.

Would it be too much work to still support building without a host Swift compiler after this issue?

@drodriguez, maybe installing a host Swift compiler on the community Android CI would be best?

I don't know the best way forward here, just pointing out the problem of #70008 breaking the build on _all_ platforms without a host Swift compiler.

drodriguez commented:
> @drodriguez, maybe installing a host Swift compiler on the community Android CI would be best?

I do not have time to work on those machines that much. I cannot also give anyone access to the machines because my company policies. Anyone is free to donate other machines that they control and can be added to the Community CI. Those machines are there as a favor, but I cannot invest that much time in them

ahoppen committed:
[SourceKit] Support building sourcekitd without building swift-syntax

This allows building sourcekitd and swift-refactor with `SWIFT_BUILD_SWIFT_SYNTAX=NO`. In these builds, the `relatedidents` and `find-syntactic-rename-ranges` requests will always return an error.

ahoppen commented:
Updated the PR so that sourcekitd and swift-refactor should build fine with `SWIFT_BUILD_SWIFT_SYNTAX=NO`. In these builds, the `relatedidents` and `find-syntactic-rename-ranges` requests will always return an error and the corresponding functionality in sourcekit-lsp will likewise not be available.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

finagolfin commented:
Thanks, this fixed that issue, but now [it is complaining about `error: cannot find -lswiftIDEUtilsBridging`](https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-16.04-android/6055/console): maybe one last tweak to make that gated on building swift-syntax too?

ahoppen commented:
Let’s see if this fixes it: https://github.com/apple/swift/pull/70269
